Utilities/Install-CustomScript.ps1

```diff
@@ -0,0 +1,177 @@
+	Initialize-Environment -AzureSecretsFile $AzureSecretsFile -logFileName "Install-CustomScriptOnAllVMs.log"
+
+	$uriArray = $FileUris -split ","
+	$settings =  @{"fileUris" = $uriArray; "commandToExecute" = $CommandToRun}
+	$protectedSettings = @{}
+	if ($StorageAccountKey -and $StorageAccountName) {
+		$protectedSettings = @{"storageAccountName" = $StorageAccountName; "storageAccountKey" = $StorageAccountKey}
+	}
+
+	if ($OSType -eq "Linux") {
+		$extensionName = "CustomScript"
+		$extensionPublisher = "Microsoft.Azure.Extensions"
+		$extensionVersion = "2.0"
+	} else {
+		$extensionName = "CustomScriptExtension"
+		$extensionPublisher = "Microsoft.Compute"
+		$extensionVersion = "1.9"
+	}
+
+	$vms = @()
+	if ($ResourceGroupName -and $VmName) {
+		$vms += Get-AzureRmVM -ResourceGroupName $ResourceGroupName -Name $VmName
+	} elseif ($ResourceGroupName) {
+		$vms = Get-AzureRmVM -ResourceGroupName $ResourceGroupName | Where-Object {$_.StorageProfile.OsDisk.OsType -eq $OSType}
+	} else {
+		$vms = Get-AzureRmVM | Where-Object {$_.StorageProfile.OsDisk.OsType -eq $OSType}
+	}
+	$jobs = @()
+	$jobIdToVM = @{}
+	$vmsNotRunning = @()
+	$vmsAgentUnresponsive = @()
+	$vmsExtensionInstalledFailed = @()
+	$alreadyInstalled = $false
+	foreach ($vm in $VMs) {
+		try {
+			$vmStatus = Get-AzureRmVM -ResourceGroupName $vm.ResourceGroupName -Name $vm.Name -Status
+			if ($vmStatus.Statuses[1].Code -inotmatch "running") {
+				$vmsNotRunning += $vm
+				continue
+			}
+			if ($vmStatus.VMAgent.VmAgentVersion -imatch "Unknown") {
+				$vmsAgentUnresponsive += $vm
+				continue
+			}
+			# Only install the extenstion on VMs running and has waagent installed
+			$extension = Get-AzureRmVMExtension -ResourceGroupName $vm.ResourceGroupName -VMName $vm.Name -Name $extensionName -ErrorAction SilentlyContinue
+			if ($extension -and $extension.PublicSettings -imatch $uriArray[0] -and $extension.PublicSettings -imatch $CommandToRun) {
+				if ($extension.ProvisioningState -eq "Failed") {
+					$vmsExtensionInstalledFailed += $vm
+				}
+				# CustomScript extension is already installed
+				if (-not $alreadyInstalled) {
+					Write-LogInfo "Custom script is already installed on the following VMs:"
+					$alreadyInstalled = $true
+				}
+				Write-LogInfo "- VM $($vm.Name) in $($vm.ResourceGroupName)"
+				continue
+			}
+
+			$job = Set-AzureRmVMExtension -ResourceGroupName $vm.ResourceGroupName -VMName $vm.Name -Location $vm.Location -Name $extensionName -Publisher $extensionPublisher `
+				-Type $extensionName -TypeHandlerVersion $extensionVersion -Settings $settings -ProtectedSettings $protectedSettings -AsJob
+			$jobs += $job
+			$jobIdToVM[$job.Id] = $vm
+		} catch {
+			Write-LogErr "Exception occurred in when installing CustomScript extension on VM $($vm.Name)."
+			Write-LogErr $_.Exception
+		}
+	}
+	$jobs | Wait-Job -Timeout 360
+	foreach ($job in $jobs) {
+		$state = $job.State
+		if ($state -eq "Running") {
+			$state = "Timeout"
```

Comment:
You did not put timeout-ed VM in the failed count. Does it mean we only count VM in power-off state if it fails?

Reply:
will include the timeout VM in the failed count.

Reply:
updated